src/Collections-Tests/WaitfreeQueueTest.class.st

```diff
@@ -0,0 +1,216 @@
+Class {
+	#name : #WaitfreeQueueTest,
+	#superclass : #TestCase,
+	#category : #'Collections-Tests-Atomic'
+}
+
+{ #category : #'helper methods' }
+WaitfreeQueueTest >> newItems: count [
+	"Answer a collection with the requested number of items"
+
+	| collection |
+
+	collection := OrderedCollection new.
+	count timesRepeat: [ collection add: AtomicQueueItem new ].
+	^collection
+]
+
+{ #category : #'helper methods' }
+WaitfreeQueueTest >> queueFromCollection: aCollection [
+	"Answer a WaitfreeQueue initialised from the supplied collection of items"
+
+	| queue |
+
+	queue := WaitfreeQueue new.
+	(queue instVarNamed: #dummy) next: aCollection first.
```

Comment:
Can we avoid instVarNamed? 
It breaks encapsulation and code navigation. No way to find those methods using var references

Reply:
In this case it should method of queue class instead of test class.

Reply:
Hi Denis,

> Can we avoid instVarNamed?
> It breaks encapsulation and code navigation. No way to find those methods using var references

I'm interested in your thoughts here.  There's two rules that both make sense:

1. Don't access internal state using `#instVarNamed:` for the reasons you outline above, as well as being slow.
2. Don't expose internal state that other objects shouldn't have access to (the dummy variable in this case).

I try to balance this by not using `#instVarNamed:` in production code, but using it for setting up internal state in associated tests (which are intimately tied to the tested class).

I can add a #dummy access method, but it will the be breaking the second rule.

Reply:
I would say let's be honest. In Smalltalk we can't really avoid exposed internal state. Because we need to examine the state in tests, we need to examine the state during navigation, debugging and scripting. Trying to do it through reflection does not change those needs. It just complicates our life. 
But it does not mean that we should design our code using message chains through accessors. We all know best practice, right?  

And here we also can avoid it by moving method to queue itself:
```Smalltalk
WaitfreeQueue withAll: aCollection
```
instead of 
```Smalltalk
WaitfreeQueueTest >> queueFromCollection: aCollection
```

Reply:
Hi Denis,

Beautifully put, obvious now that you've said it, and I'm feeling embarrassed that you had to say it.

I've pushed an update.

Thanks!
Alistair